### PR TITLE
[v7r1] First steps towards json serialization

### PR DESCRIPTION
**Pull request by chaen**
This PR contains the changes needed to move to JSON.
Basically, it makes the necessary casts explicitly (see https://github.com/DIRACGrid/DIRAC/issues/3551) 

Since v7r1 was broken when I started that, I've based the work on v7r0 for the time being

Integration tests ran here: https://jenkins-dirac.web.cern.ch/view/Development/job/jsonSerialization/

BEGINRELEASENOTES

*Subsystem
NEW/CHANGE/FIX: explanation

For examples look into release.notes

ENDRELEASENOTES

### CODE REVIEW

**Comment by chrisburr** [comment]
As this will be included in v7r1 and Chris is away this week should this be merged in advance of the hackathon tomorrow?

I've ran the integration tests for this against `rel-v7r1` and #4289 and [everything looks okay](https://github.com/chrisburr/DIRAC/pull/3/checks?check_run_id=244425922).

**Comment by atsareg** [comment]
Can it be rebased on rel-v7r1 ? Also there are some formatting problems

**Comment by fstagni** [comment]
I have just set the target to v7r1, and there's a conflict which looks pretty easy to solve.

File: Core/DISET/TransferClient.py [hunk 1]
```diff
@@ -31,7 +31,8 @@ def _sendTransferHeader( self, actionName, fileInfo ):
       retVal = self._proposeAction( transport, ( "FileTransfer", actionName ) )
       if not retVal[ 'OK' ]:
         return retVal
-      retVal = transport.sendData( S_OK( fileInfo ) )
+      # We need to convert to list
+      retVal = transport.sendData( S_OK( list(fileInfo) ) )
```
Comment (fstagni) [inline Core/DISET/TransferClient.py]:
```suggestion
      retVal = transport.sendData(S_OK(list(fileInfo)))
```

File: Core/DISET/private/FileHelper.py [hunk 2]
```diff
@@ -54,14 +54,14 @@ def getTransferedBytes( self ):
...
-    retVal = self.oTransport.sendData( S_OK( ( True, sBuffer ) ) )
+    retVal = self.oTransport.sendData( S_OK( [ True, sBuffer ] ) )
     if not retVal[ 'OK' ]:
       return retVal
     retVal = self.oTransport.receiveData()
     return retVal
 
   def sendEOF( self ):
-    retVal = self.oTransport.sendData( S_OK( ( False, self.__oMD5.hexdigest() ) ) )
+    retVal = self.oTransport.sendData( S_OK( [ False, self.__oMD5.hexdigest() ] ) )
```
Comment (fstagni) [inline Core/DISET/private/FileHelper.py]:
```suggestion
    retVal = self.oTransport.sendData(S_OK([False, self.__oMD5.hexdigest()]))
```

File: Core/DISET/private/InnerRPCClient.py [hunk 3]
```diff
@@ -58,7 +58,10 @@ def executeRPC( self, functionName, args ):
             return retVal
 
       # Send the arguments to the function
-      retVal = transport.sendData( S_OK( args ) )
+      # Note: we need to convert the arguments to list
+      # We do not need to deseralize it because variadic functions
+      # can work with list too
+      retVal = transport.sendData( S_OK( list(args) ) )
```
Comment (fstagni) [inline Core/DISET/private/InnerRPCClient.py]:
```suggestion
      retVal = transport.sendData(S_OK(list(args)))
```

**Comment by fstagni** [comment]
I solved the conflict. The style issues I think are highlighted above.

File: Core/Utilities/DEncode.py [hunk 4]
```diff
@@ -366,8 +373,7 @@ def decodeDict(data, i):
     if DIRAC_DEBUG_DENCODE_CALLSTACK:
       # If we have numbers as keys
       if data[i] in ('i', 'I', 'f'):
-        print('=' * 40, "Decoding dict with numeric keys", '=' * 40)
-        printDebugCallstack()
+        printDebugCallstack( "Decoding dict with numeric keys")
```
Comment (andresailer) [inline Core/Utilities/DEncode.py]:
```suggestion
        printDebugCallstack("Decoding dict with numeric keys")
```

**Comment by chaen** [comment]
It needs more rework than just rebase as Andrei added the client version in the argument. Are the jenkins integration tests of v7r1 now working completely ?

**Comment by chrisburr** [comment]
Yes they are

**Pull request closed without merging**